open_rarity/models/collections/_utils.py

```diff
@@ -0,0 +1,163 @@
+from collections import defaultdict
+from itertools import chain
+
+from satchel import groupapply
+
+from open_rarity.models.tokens import (
+    AttributeName,
+    RawToken,
+    TokenIdMetadataAttr,
+    TokenSchema,
+)
+
+
+def flatten_token_data(tokens: list[RawToken]) -> list[TokenIdMetadataAttr]:
+    """Denormalized and flatten token data. Attributes move to the top level and are
+    assign a `token_id: <id>` key.
+
+    Parameters
+    ----------
+    tokens : list[Token]
+        _description_
+
+    Returns
+    -------
+    list[TokenIdMetadataAttr]
+        _description_
+    """
+    return list(
+        chain(
+            *[
+                [{"token_id": t["token_id"], **attr} for attr in t["attributes"]]
+                for t in tokens
+            ]
+        )
+    )
+
+
+def extract_token_name_key(t: TokenIdMetadataAttr) -> tuple[int | str, str]:
+    return t["token_id"], t["name"]
+
+
+def _create_token_schema(tokens: list[TokenIdMetadataAttr]) -> TokenSchema:
+    """Create a schema that is representative of a token containing all possible
+    attribute keys and the correct number of them.
+
+    Parameters
+    ----------
+    tokens : list[TokenIdMetadataAttr]
+        _description_
+
+    Returns
+    -------
+    TokenSchema
+        _description_
+    """
+    d = defaultdict(int)
+    for key, count in groupapply(tokens, extract_token_name_key, "count").items():
```

Comment:
Not in favor of additional dependencies - can we have a list of dependencies and discuss what's needed and what's not?